### PR TITLE
Add SIMD using OpenMP

### PR DESCRIPTION
**Pull request by MatthieuHernandez**

### CODE REVIEW

File: src/neural_network/layer/neuron/SimpleNeuron.cpp [hunk 1]
```diff
@@ -15,19 +15,21 @@ SimpleNeuron::SimpleNeuron(NeuronModel model, shared_ptr<NeuralNetworkOptimizer>
...
-    this->sum += bias;
+    this->sum = tmp + bias;
     return this->outputFunction->function(this->sum);
 }
 
 vector<float>& SimpleNeuron::backOutput(float error)
 {
     error = error * this->outputFunction->derivative(this->sum);
 
+    #pragma omp simd // seems to do nothing
```
Comment (andreasceid) [inline src/neural_network/layer/neuron/SimpleNeuron.cpp:32]: Replace `#pragma omp simd` with `#pragma omp parallel for num_threads(/*number of logical cores*/) schedule(runtime)`

File: src/neural_network/layer/neuron/RecurrentNeuron.cpp [hunk 2]
```diff
@@ -22,12 +22,13 @@ float RecurrentNeuron::output(const vector<float>& inputs, bool temporalReset)
     this->previousOutput = this->lastOutput;
     this->sum = 0;
     int w;
+    float tmp = 0.0f; // to activate the SIMD optimization
+    //#pragma omp simd
```
Comment (andreasceid) [inline src/neural_network/layer/neuron/RecurrentNeuron.cpp:26]: Uncomment SIMD routine

File: src/neural_network/layer/neuron/RecurrentNeuron.cpp [hunk 3]
```diff
@@ -37,6 +38,7 @@ std::vector<float>& RecurrentNeuron::backOutput(float error)
 {
     error = error * outputFunction->derivative(this->sum);
 
+    //#pragma omp simd // seems to do nothing
```
Comment (andreasceid) [inline src/neural_network/layer/neuron/RecurrentNeuron.cpp:41]: Replace `#pragma omp simd` with `#pragma omp parallel for num_threads(/*number of logical cores*/) schedule(runtime)`

**Review by andreasceid — COMMENTED** [review]
I have recommended some changes on your OpenMP routines. There is a chance that some of them do not directly compile, because of `size_t` variables in the for loops. If that is the case, typecast those variables into integers and retry. 

Keep me posted :)

**Comment by MatthieuHernandez** [comment]
When I compile with MSVC I have an info message if the for loop isn't vectorized:
`info C5002: Omp simd loop not vectorized due to reason '1200'`
So here all the loop using OPenMP is well vectoried. 
An `size_t` is just a `unsigned long long` the size of an memory address. 
I don't kwon why an `int` would be better. Can you show me a docment when it write it's better to use an `int` ?

File: src/neural_network/layer/neuron/SimpleNeuron.cpp (hunk 1 above)
Comment (MatthieuHernandez) [reply src/neural_network/layer/neuron/SimpleNeuron.cpp:32]: I have to think if I want to set multi-core in this branch, if so I have to think about how to configure the number of cores to use in the library.

File: src/neural_network/layer/neuron/RecurrentNeuron.cpp (hunk 3 above)
Comment (MatthieuHernandez) [reply src/neural_network/layer/neuron/RecurrentNeuron.cpp]: I have to think if I want to set multi-core in this branch, if so I have to think about how to configure the number of cores to use in the library.

File: src/neural_network/layer/neuron/RecurrentNeuron.cpp (hunk 2 above)
Comment (MatthieuHernandez) [reply src/neural_network/layer/neuron/RecurrentNeuron.cpp]: It's a mistake, I uncommented it.

**Comment by andreasceid** [comment]
To find the number of logical cores of the host system, try:

- For Unix: `cat /proc/cpuinfo | grep siblings | uniq`
- For Windows: `Get-WmiObject -class win32_processor -Property "NumberOfLogicalProcessors" | Select-Object -Property "NumberOfLogicalProcessors"`

**Review by andreasceid — APPROVED** [review]
Check any compiler flags that can further boost performance. For example `-O3` for level 3 optimizations.

**Pull request merged**